### PR TITLE
Finish the Multiple Choice Classification

### PR DESCRIPTION
**Pull request by oberonbot**
Run through the whole process, and hard code the dataset.

### CODE REVIEW

File: test/nlp/test_autohf_multichoice_classification.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    true_count = 0
+    for i, v in X_true.items():
+        if y_pred[i] == v:
+            true_count += 1
+    accuracy = round(true_count / len(y_pred), 5)
+    print("Accuracy: " + str(accuracy))
+
+
+if __name__ == "__main__":
+    test_hf_data()
```
Comment (liususan091219) [inline test/nlp/test_autohf_multichoice_classification.py:208]: rename test_hf_data -> test_mcc

File: test/nlp/test_autohf_multichoice_classification.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
+import os
+import pytest
+
+
+@pytest.mark.skipif(os.name == "posix", reason="do not run on mac os")
+def test_hf_data():
+    from flaml import AutoML
```
Comment (liususan091219) [inline test/nlp/test_autohf_multichoice_classification.py:7]: Remove all comments

File: test/nlp/test_autohf_multichoice_classification.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
+import os
+import pytest
+
+
+@pytest.mark.skipif(os.name == "posix", reason="do not run on mac os")
```
Comment (liususan091219) [inline test/nlp/test_autohf_multichoice_classification.py:5]: Should be "darwin" not "posix"

**Review by liususan091219 — CHANGES_REQUESTED** [review]

**Review by liususan091219 — APPROVED** [review]

File: flaml/model.py [hunk 4]
```diff
@@ -667,6 +675,8 @@ def predict_proba(self, X_test):
         from .nlp.huggingface.trainer import TrainerForAuto
         from transformers import TrainingArguments
         from .nlp.utils import load_model
+        from transformers import AutoTokenizer
```
Comment (sonichi) [inline flaml/model.py:678]:
```suggestion
        from transformers import TrainingArguments, AutoTokenizer
        from .nlp.utils import load_model
```

File: flaml/model.py [hunk 5]
```diff
@@ -704,7 +728,19 @@ def predict(self, X_test):
...
+        self._model = TrainerForAuto(
+            model=best_model,
+            args=training_args,
+            data_collator=DataCollatorForPredict(
+                tokenizer=tokenizer,
+                pad_to_multiple_of=8 if training_args.fp16 else None,
+            )
+            if self._task == MULTICHOICECLASSIFICATION
+            else None,
+        )
```
Comment (sonichi) [inline flaml/model.py:743]: can you merge the identical code in predict_proba() and predict() ?

File: flaml/nlp/utils.py [hunk 6]
```diff
@@ -140,6 +152,64 @@ def tokenize_row(
     return [tokenized_example[x] for x in tokenized_column_names]
 
 
+def tokenize_text_multiplechoice(X, custom_hpo_args):
+    from transformers import AutoTokenizer
+    import pandas
+
+    global tokenized_column_names
```
Comment (sonichi) [inline flaml/nlp/utils.py:159]: don't use global variables.

File: flaml/nlp/utils.py [hunk 7]
```diff
@@ -140,6 +152,64 @@ def tokenize_row(
...
+    # get each 1st sentence, multiply to 4 sentences
+    question_headers = this_row["sent2"]
+    # sent2 are the noun part of 2nd line
+    second_sentences = [
+        question_headers + " " + i
+        for i in [
+            this_row["ending0"],
+            this_row["ending1"],
+            this_row["ending2"],
+            this_row["ending3"],
```
Comment (sonichi) [inline flaml/nlp/utils.py:194]:
```suggestion
        question_headers + " " + this_row[key]
        for key in ["ending0", "ending1", "ending2", "ending3"]
```

File: flaml/nlp/utils.py [hunk 8]
```diff
@@ -358,7 +437,7 @@ class HPOArgs:
         metadata={"help": "model path model for HPO"},
     )
 
-    fp16: bool = field(default=True, metadata={"help": "whether to use the FP16 mode"})
+    fp16: bool = field(default=False, metadata={"help": "whether to use the FP16 mode"})
```
Comment (sonichi) [inline flaml/nlp/utils.py:440]: This seems a big change. @liususan091219 could you confirm this is desired?

File: test/nlp/test_autohf_multichoice_classification.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    y_val = dev_dataset[label_key]
+
+    X_test = test_dataset[custom_sent_keys]
+    X_true = test_dataset[label_key]
+    automl = AutoML()
+
+    automl_settings = {
+        "gpu_per_trial": 0,
+        "max_iter": 2,
+        "time_budget": 10000,
```
Comment (sonichi) [inline test/nlp/test_autohf_multichoice_classification.py:142]: too long.

**Review by sonichi — CHANGES_REQUESTED** [review]
Please fix the lint issues. You can use [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to help you.

File: flaml/nlp/utils.py (hunk 8 above)
Comment (liususan091219) [reply flaml/nlp/utils.py:440]: This is on purpose. If we set it to True all the test should explicitly specify fp16. There are a lot of tests so I decided the default should be false.

File: test/nlp/test_autohf_multichoice_classification.py (hunk 9 above)
Comment (liususan091219) [reply test/nlp/test_autohf_multichoice_classification.py:142]:
```suggestion
        "time_budget": 5,
```

File: flaml/nlp/utils.py (hunk 8 above)
Comment (liususan091219) [reply flaml/nlp/utils.py:440]:
```suggestion
    fp16: bool = field(default=True, metadata={"help": "whether to use the FP16 mode"})
```

File: flaml/nlp/utils.py (hunk 6 above)
Comment (liususan091219) [reply flaml/nlp/utils.py:159]: We are not fixing it in this PR.

File: flaml/model.py [hunk 10]
```diff
@@ -658,53 +666,52 @@ def _compute_metrics_by_dataset_name(self, eval_pred):
             )
             return metric_dict
 
-    def predict_proba(self, X_test):
-        assert (
-            self._task in CLASSIFICATION
-        ), "predict_proba() only for classification tasks."
-
+    def init_model_for_predict(self, X_test):
```
Comment (sonichi) [inline flaml/model.py:669]: Is this method supposed to be called by external code? if not, use "_" in the beginning of the method name.

**Review by sonichi — COMMENTED** [review]
Please fix the lint issues as listed in https://github.com/microsoft/FLAML/runs/4686449608?check_suite_focus=true
Or, create an issue and have it assigned.

**Review by sonichi — APPROVED** [review]

**Pull request merged**